pkg/pathutils/pathutils.go

```diff
@@ -0,0 +1,66 @@
+/*
+Copyright © 2021 Aspect Build Systems Inc
+Not licensed for re-use.
+*/
+
+package pathutils
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+)
+
+func IsValidFile(path string) bool {
+	info, err := os.Stat(path)
+	if err != nil {
+		return false
+	}
+
+	return !info.IsDir()
+}
+
+// isValidWorkspace returns true iff the supplied path is the workspace root, defined by the presence of
+// a file named WORKSPACE or WORKSPACE.bazel
+// see https://github.com/bazelbuild/bazel/blob/8346ea4cfdd9fbd170d51a528fee26f912dad2d5/src/main/cpp/workspace_layout.cc#L37
+func IsValidWorkspace(path string) bool {
+	return IsValidFile(filepath.Join(path, "WORKSPACE")) ||
+		IsValidFile(filepath.Join(path, "WORKSPACE.bazel"))
+}
+
+// IsValidPackage returns true iff a file named BUILD or BUILD.bazel exists
+// within the dir at the specified path
+func IsValidPackage(path string) bool {
+	return IsValidFile(filepath.Join(path, "BUILD")) ||
+		IsValidFile(filepath.Join(path, "BUILD.bazel"))
+}
+
+func FindWorkspaceRoot(path string) string {
+	if IsValidWorkspace(path) {
+		return path
+	}
+
+	parentDirectory := filepath.Dir(path)
+	if parentDirectory == path {
+		return ""
+	}
```

Comment:
I don't really understand this check. It would only be true if the path is relative, and it is `.`. Otherwise, if I'm reading correctly, the recursive function will run forever if it's not a Bazel workspace. Combined with the growing stack, you will run out of memory.

Reply:
No, this actually works correctly. This stopping condition occurs when we've reached the root directory on disk, ie. when the current folder's parent is itself.

Reply:
`filepath.Dir(path)` gets the parent directory of `path`. I'll add a comment describing this.